### PR TITLE
Build on zio-aws

### PR DESCRIPTION
**Pull request by svroonland**
`zio-aws` offers ZIO wrappers over the (low-level) AWS SDK methods, leaving `zio-sqs` to provide higher level functionality like streaming consuming and a producer.

One benefit of this change would be a nicely integrated ecosystem of compatible libraries around various AWS services.

I added code files for the examples in the readme to check compilation.

### CODE REVIEW

**Comment by gchudnov** [comment]
@svroonland 
Hi,

a bit worried:
- about the correctness of https://github.com/vigoo/zio-aws, given that all code was auto-generated. I see almost no tests there.
- readability of *zio-aws-sqs* looks not very good. I see no sources included for the auto-generated code (if we open sources [here](https://repo1.maven.org/maven2/io/github/vigoo/zio-aws-sqs_2.13/3.15.19.8/), there is literally nothing inside.)

in this way, for example when trying to get what is the signature of `SendMessageBatchRequestEntry`, I can see only:

image

with lots of compiled code with an option to decompile to Java.

If for some reason there is a bug in *zio-aws-sqs* it might be a nightmare to debug it.

--
but not being against in general (and I'm not the creator of zio-sqs :) )

**Comment by svroonland** [comment]
Fair concerns @gchudnov. 

Regarding tests, I believe zio-aws has quite some tests on the level of types of SDK methods, so a test suite for methods that return a stream for example. Keep in mind that the AWS java SDK itself is code generated as well. Also, the test suite of `zio-sqs` covers the behavior used from `zio-aws-sqs` of course, so overall test coverage for `zio-sqs` has not decreased. 

For the compiled sources, perhaps there is a way to publish those along with the binaries, I'd have to check with the zio-aws maintainer.

Anyways, I was perhaps a bit quick to submit a PR for this, but it wasn't an awful lot of work to change it and it's good to have feedback and discussion on it if this is a good direction for zio-aws.

**Comment by ghostdogpr** [comment]
Good points. I am not too concerned about the first issue since `zio-aws` is not doing anything complex or custom but wrapping the AWS SDK in a generic way. The fact it's automated feels even safer to me 😄 

The second point is more annoying, it would be nice if there was a solution.

By the way, I am no longer using SQS at work so I am not very involved in this (plus I have to maintain a bunch of other stuff), so if you guys are interested in becoming official maintainers that could be awesome. Please let me know!

**Comment by vigoo** [comment]
Hi all, I'm the author of `zio-aws`, some remarks to your concerns:

- About testing. All the operations in the AWS SDK match one of the operation types that are implemented in a generic way here: https://github.com/vigoo/zio-aws/blob/master/zio-aws-core/src/main/scala/io/github/vigoo/zioaws/core/AwsServiceBase.scala. For SQS it means all operations are wrapped with either `asyncRequestResponse` or `asyncSimplePaginatedRequest`. There are tests for all these core wrapper functions, trying to simulate all possible failure cases.

The main advantage of this approach in my opinion is that you don't have to test against the actual services. We can delegate that to the authors of the AWS Java SDK and if we trust that library _and_ the tests for these few core wrapper functions, we immediately covered the full wrapped SDK.

- In addition to the above I included a few integration tests that test some of the generated libraries against localstack. These are more useful to test the non-netty backends though, and also serve as simple usage examples. We can add SQS tests to that suite if it would give more confidence that it works.

- I did not notice that the generated sources are not published, will try to figure out why. Until then, for the sake of discussion I copied the generated code for `zio-aws-sqs` to this gist: https://gist.github.com/vigoo/e5ab050b8b01baaaf77ef99e00ffaf7b

**Comment by gchudnov** [comment]
@vigoo @svroonland 
got it, thank you for the clarification.

@ghostdogpr 
not having much time myself, I'm still interested in contributing a streaming consumer (+ probably polishing producer a bit).
Believe I still have the code I was preparing for the PR last year.

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (gchudnov) [inline src/main/scala/zio/sqs/producer/Producer.scala:181]: is it worth to have a named parameters here?
like:
```scala
        SendMessageBatchRequestEntry(
          id = index.toString,
          messageBody = serializer(e.event.data),
...
```
just to improve the readability?

since in the prev version it was more readable with the builder pattern, that is removed now, and looking at 
```scala
SendMessageBatchRequestEntry(
...
None,
...
```
for example, it is not entirely clear what it referes to

and the same comment regarding the construction of `ReceiveMessageRequest` above.

File: src/main/scala/zio/sqs/SqsStream.scala
Comment (gchudnov) [inline src/main/scala/zio/sqs/SqsStream.scala:21]: hm.. is `map(_.toInt),` required?
it looks the type is already integer.

--
and actually looking at the settings now, is it worth to replace with integer durations with `zio.duration` type?
but might be worth to make in a separate PR, might be out of scope for this one.

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (gchudnov) [inline src/main/scala/zio/sqs/producer/Producer.scala:201]: io.github.vigoo.zioaws.sqs
^
sorry for the stupid question, any particular reason whey `io.github.vigoo.zioaws.sqs` is used explicitly without importing it instead like:

```scala
import io.github.vigoo.zioaws.sqs._
```

and then:
```scala
...
  ): RIO[R with Clock with Sqs, Unit] =
      sendMessageBatch(req.inner)
      .mapError(_.toThrowable)
...
```

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (svroonland) [reply src/main/scala/zio/sqs/producer/Producer.scala:181]: Definitely, I was just working through the changes as quickly as possible to get things to compile and test. Now is the time for cleaning up.

File: src/main/scala/zio/sqs/SqsStream.scala
Comment (svroonland) [reply src/main/scala/zio/sqs/SqsStream.scala:21]: One is a scala int, the other is a java one. 

Agreed about zio.duration, but I wanted to keep changes minimal in this PR.

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (svroonland) [reply src/main/scala/zio/sqs/producer/Producer.scala:201]: Not at all a stupid question. I wanted to add some sort of qualifer to the method instead of just `sendMessageBatch` (where does it come from?). But `sqs.sendMessageBatch(req.inner)` doesn't compile since it looks for `zio.sqs.sendMessageBatch` namespace. Later on I used `import io.github.vigoo.zioaws` so you can do `zioaws.sqs.sendMessageBatch`

File: src/main/scala/zio/sqs/SqsStream.scala
Comment (svroonland) [reply src/main/scala/zio/sqs/SqsStream.scala:21]: Could the java/scala Integer/int thing be addressed in `zio-aws` @vigoo?

Comment (vigoo) [reply src/main/scala/zio/sqs/SqsStream.scala]: The goal is that on the zio-aws interfaces everything is Scala friendly, including all primitive and collection types. If in this case you have to pass a boxed java `Integer` that's something that should be fixed in `zio-aws`. I'll take a look tomorrow.

Comment (vigoo) [reply src/main/scala/zio/sqs/SqsStream.scala]: Integer fix is going to be published soon: https://github.com/vigoo/zio-aws/pull/120

**Comment by svroonland** [comment]
Alright, with the updated readme, this one is up for review. If this is something you guys like, it might be good to wait for a bit before merging/releasing, I know @vigoo is planning a microsite with some more docs on zio-aws.

File: README.md
Comment (gchudnov) [inline README.md:176]: just wondering: there is no zio-aws wrapped for this builder, right?

**Review by gchudnov — APPROVED** [review]
only one minor comment regarding the builder

in general looks fine :+1:

File: README.md
Comment (vigoo) [reply README.md:176]: The SQS client builder does not have any SQS-specific options so instead of this the configuration could come from the `AwsConfig` layer. It has a `zio-config` based layer definition, with that it could be:

```scala
ZLayer.success(CommonAwsConfig(region = Some(Region.of("ap-northeast-2")), credentialsProvider = ..., endpointOverride = None, commonClientConfig = None)) >>> zioaws.core.config.configured() >>> zioaws.sqs.live
```

Comment (vigoo) [reply README.md:176]: The idea of the `AwsConfig` layer is that it:
- gets the http backend as a dependency and applies it to the AWS client builder
- provides all the common settings for the client builder

There are a few AWS clients that has some custom options too, and that's not desribed by the schema. Those custom properties currently can only be set with the `servicelayer.customized(builderFn)` way as seen above. I was thinking of reverse engineering them from the code generator with reflection but there are so few of them that I think it does not worth it.

File: build.sbt
Comment (ghostdogpr) [inline build.sbt:28]: I realize a good aspect of that is I won't receive a Scala Steward PR daily to upgrade the AWS library 😆

File: src/main/scala/zio/sqs/Utils.scala
Comment (ghostdogpr) [inline src/main/scala/zio/sqs/Utils.scala:12]: It's a small thing but hot about using the shorter aliases like `RIO`?

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (ghostdogpr) [inline src/main/scala/zio/sqs/producer/Producer.scala:220]: Nice optimization using `foreach_` 👍

File: src/test/scala/zio/sqs/ZioSqsMockServer.scala
Comment (ghostdogpr) [inline src/test/scala/zio/sqs/ZioSqsMockServer.scala:19]: You're right, the `Task` is useless here I think

File: src/main/scala/zio/sqs/producer/Producer.scala
Comment (svroonland) [reply src/main/scala/zio/sqs/producer/Producer.scala:220]: Thanks to the zio-intellij plugin! :)

File: build.sbt
Comment (svroonland) [reply build.sbt:28]: There might be those for zio-aws itself. I am please with `mergify` for automatically merging those PRs, some other zio projects are using it too.

**Comment by gchudnov** [comment]
@ghostdogpr @svroonland 
anything left in this PR to do? If not, let's merge it.

**Review by ghostdogpr — APPROVED** [review]
All good with me, I think the improvements from zio-aws are included. 

@svroonland OK to merge this?

**Comment by svroonland** [comment]
Good to go for me!

**Pull request merged**
